### PR TITLE
[WIP] Vecsim hybrid query

### PR DESCRIPTION
**Pull request by alonre24**
Replace the vector "list iterator" with a "hybrid iterator" that allows hybrid queries of rediSearch with vector similarity search. Implementation is based on the following design: https://redislabs.atlassian.net/wiki/spaces/DX/pages/3237478445/FT.SEARCH+Execution

### CODE REVIEW

File: src/hybrid_reader.c [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+      *upper_bound = VECTOR_RESULT(top)->dist.distance;
+      // Reset the current result and advance both "sub-iterators".
+      AggregateResult_Reset(cur_res);
+      hr->childIt->Read(hr->childIt->ctx, &cur_child_res);
+      HR_ReadInBatch(hr, &cur_vec_res);
+    }
+    // Otherwise, advance the iterator pointing to the lower id.
+    else if (cur_vec_res->docId > cur_child_res->docId && hr->childIt->isValid) {
+      int rc = hr->childIt->SkipTo(hr->childIt->ctx, cur_vec_res->docId, &cur_child_res);
+      if (rc == INDEXREAD_NOTFOUND) {
```
Comment (alonre24) [inline src/hybrid_reader.c:138]: Add clarification

File: src/hybrid_reader.c [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+      // consider binary search for next value
+      continue;
+    }
+    // Set the item that we skipped to it in hit.
+    (*hit)->docId = VecSimQueryResult_GetId(res);
+    (*hit)->dist.distance = VecSimQueryResult_GetScore(res);
+    (*hit)->dist.scoreField = hr->scoreField;
+    return INDEXREAD_OK;
+  }
+  if (hr->mode == STANDARD_KNN) {
```
Comment (alonre24) [inline src/hybrid_reader.c:71]: Remove this

File: src/aggregate/aggregate_request.c [hunk 3]
```diff
@@ -928,15 +928,34 @@ static ResultProcessor *getGroupRP(AREQ *req, PLN_GroupStep *gstp, ResultProcess
   return pushRP(req, groupRP, rpUpstream);
 }
 
-#define _SCORE_LEN 6
+static ResultProcessor *getVecSimRP(AREQ *req, AGGPlan *pln, ResultProcessor *up, QueryError *status) {
+  RLookup *lk = AGPLN_GetLookup(pln, NULL, AGPLN_GETLOOKUP_LAST);
+  RLookupKey **keys = rm_calloc(array_len(req->ast.vecScoreFieldNames), sizeof(*keys));
+
```
Comment (ashtul) [inline src/aggregate/aggregate_request.c:934]: add `char **ptr = req->ast.vecScoreFieldNames` pointer for readibility

File: src/hybrid_reader.c [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+} VecSearchMode;
+
+typedef struct {
+  IndexIterator base;
+  VecSimIndex *index;
+  TopKVectorQuery query;
+  VecSimQueryParams runtimeParams;   // evaluated runtime params.
+  IndexIterator *childIt;
+  VecSearchMode mode;
+  bool RESULTS_PREPARED;
```
Comment (ashtul) [inline src/hybrid_reader.c:19]: why is it capitilized?

File: src/hybrid_reader.c [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+  VecSimQueryResult_List list;
+  VecSimQueryResult_Iterator *iter;
+  t_docId lastDocId;
+  RSIndexResult **returnedResults; // Save the pointers to be freed in clean-up.
+  char *scoreField;  // to use by the sorter, for distinguishing between different vector fields.
```
Comment (ashtul) [inline src/hybrid_reader.c:24]: not used

File: src/hybrid_reader.c [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+  VecSimQueryResult *res = VecSimQueryResult_IteratorNext(hr->iter);
+  // Set the item that we read in the current RSIndexResult
+  (*hit)->docId = VecSimQueryResult_GetId(res);
+  (*hit)->dist.distance = VecSimQueryResult_GetScore(res);
+  (*hit)->dist.scoreField = hr->scoreField;
+  return INDEXREAD_OK;
+}
+
+static void alternatingIterate(HybridIterator *hr, VecSimQueryResult_Iterator *vecsim_iter, float *upper_bound) {
+  RSIndexResult *cur_vec_res = NewDistanceResult(), *cur_child_res, *cur_res = hr->base.current;
```
Comment (ashtul) [inline src/hybrid_reader.c:102]: break into multiple lines

File: src/hybrid_reader.c [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+  if (!hr->base.isValid) return false;
+  if (hr->mode == STANDARD_KNN && hr->RESULTS_PREPARED) {
+    return VecSimQueryResult_IteratorHasNext(hr->iter);
+  }
+  if (hr->RESULTS_PREPARED) {
+    return heap_count(hr->topResults) > 0;
+  }
+  // Otherwise, this is called before we prepared the results. There is at least one result
+  // if the child_it has at least one result.
+  return hr->childIt->Len(hr->childIt) > 0;
```
Comment (ashtul) [inline src/hybrid_reader.c:258]: consider using `IITER_HAS_NEXT`

**Review by ashtul — COMMENTED** [review]
Looks great

File: src/hybrid_reader.c [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+
+static size_t HR_Len(void *ctx) {
+  HybridIterator *hr = ctx;
+  if (!hr->RESULTS_PREPARED) {
+    prepareResults(hr);
+  }
+  if (hr->mode == STANDARD_KNN) {
+    return VecSimQueryResult_Len(hr->list);
+  }
+  return heap_count(hr->topResults);
```
Comment (ashtul) [inline src/hybrid_reader.c:224]: return `k`

File: src/hybrid_reader.c (hunk 7 above)
Comment (ashtul) [reply src/hybrid_reader.c:255]: return isValid

**Review by ashtul — COMMENTED** [review]
some more

File: src/query.c [hunk 9]
```diff
@@ -638,13 +638,14 @@ static IndexIterator *Query_EvalVectorNode(QueryEvalCtx *q, QueryNode *qn) {
     return NULL;
   }
   // Add the score field name to the ast score field names array.
-  // This macro creats the array if it's the first name, and ensure its size is sufficient.
+  // This macro creates the array if it's the first name, and ensure its size is sufficient.
   array_ensure_append_1(*q->vecScoreFieldNamesP, qn->vn.vq->scoreField);
+  IndexIterator *child_it = NULL;
   if (QueryNode_NumChildren(qn) > 0) {
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/query.c:644]: Yes, if the query node was generated from the rule `*=>query ::= star ARROW LSQB vector_query(B) RSQB`

File: src/redisearch.h [hunk 10]
```diff
@@ -247,16 +247,24 @@ typedef struct {
   double value;
 } RSNumericRecord;
 
+/* A vector record represents a vector similarity search result which has a specific *distance* from the
+ * query vector in the vector index, and associated with *scoreField* */
+typedef struct {
+  double distance;
+  char *scoreField;
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/redisearch.h:254]: Currently no, it a preparation for an air-conditioner.... 
In the future, if we have several vector scores in the query, we would like to know the specific vector field that this particular `RSIndexResult` is associated with.

File: src/vector_index.c [hunk 11]
```diff
@@ -117,14 +118,40 @@ IndexIterator *NewVectorIterator(RedisSearchCtx *ctx, VectorQuery *vq, QueryErro
         QueryError_SetError(status, QUERY_EINVAL, "Error parsing vector similarity query: query vector does not match index's type or dimention.");
         return NULL;
       }
-      vq->results = VecSimIndex_TopKQuery(vecsim, vq->topk.vector, vq->topk.k, &qParams, vq->topk.order );
-      vq->resultsLen = VecSimQueryResult_Len(vq->results);
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/vector_index.c:121]: This has moved to the `prepareResults` function (called in execution time)

File: src/vector_index.c [hunk 12]
```diff
@@ -117,14 +118,40 @@ IndexIterator *NewVectorIterator(RedisSearchCtx *ctx, VectorQuery *vq, QueryErro
...
 
+IndexIterator *NewHybridVectorIterator(RedisSearchCtx *ctx, VectorQuery *vq, QueryError *status, IndexIterator *child_it) {
+  RedisModuleString *key = RedisModule_CreateStringPrintf(ctx->redisCtx, "%s", vq->property);
+  VecSimIndex *vecsim = openVectorKeysDict(ctx, key, 0);
+  RedisModule_FreeString(ctx->redisCtx, key);
+  if (!vecsim) {
+    return NULL;
+  }
+  switch (vq->type) {
+    case VECSIM_QT_TOPK:;
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/vector_index.c:106]: cannot, otherwise it will not compile

File: src/hybrid_reader.c (hunk 5 above)
Comment (alonre24) [reply src/hybrid_reader.c:24]: This is on purpose (as I explained in previous comment)

File: src/hybrid_reader.c (hunk 4 above)
Comment (alonre24) [reply src/hybrid_reader.c]: changed

File: src/hybrid_reader.c [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+
+static int cmpVecSimResByScore(const void *p1, const void *p2, const void *udata) {
+  const RSIndexResult *e1 = p1, *e2 = p2;
+  double score1 = VECTOR_RESULT(e1)->dist.distance, score2 = VECTOR_RESULT(e2)->dist.distance;
+  if (score1 < score2) {
+    return -1;
+  } else if (score1 > score2) {
+    return 1;
+  }
+  return 0;
```
Comment (DvirDukhan) [inline src/hybrid_reader.c:18]: do we need a tie-breaker here?

File: src/hybrid_reader.c [hunk 14]
```diff
@@ -0,0 +1,299 @@
...
+  HybridIterator *hr = ctx;
+  while(VecSimQueryResult_IteratorHasNext(hr->iter)) {
+    VecSimQueryResult *res = VecSimQueryResult_IteratorNext(hr->iter);
+    t_docId id = VecSimQueryResult_GetId(res);
+    if (docId > id) {
+      // consider binary search for next value
+      continue;
+    }
+    // Set the item that we skipped to it in hit.
+    (*hit)->docId = VecSimQueryResult_GetId(res);
```
Comment (DvirDukhan) [inline src/hybrid_reader.c:44]:
```suggestion
    (*hit)->docId = id;
```

File: src/hybrid_reader.c (hunk 1 above)
Comment (DvirDukhan) [reply src/hybrid_reader.c:101]: please do :)

File: src/hybrid_reader.c [hunk 15]
```diff
@@ -0,0 +1,299 @@
...
+    hr->returnedResults = array_append(hr->returnedResults, *hit);
+  }
+  hr->lastDocId = (*hit)->docId;
+  return INDEXREAD_OK;
+
+eof:
+  hr->base.isValid = false;
+  return INDEXREAD_EOF;
+}
+
```
Comment (DvirDukhan) [inline src/hybrid_reader.c:183]: please split this into functions for each mode

File: src/hybrid_reader.c [hunk 16]
```diff
@@ -0,0 +1,299 @@
...
+static bool UseBF(size_t T, TopKVectorQuery query, VecSimIndex *index) {
+  // todo: have more sophisticated heuristics here, currently this is disabled.
+  // return (float)T < (0.05 * (float)VecSimIndex_IndexSize(index));
+  return false;
+}
+
+static size_t HR_NumEstimated(void *ctx) {
```
Comment (DvirDukhan) [inline src/hybrid_reader.c:190]: do we test this?

File: src/hybrid_reader.c [hunk 17]
```diff
@@ -0,0 +1,299 @@
...
+  return hr->lastDocId;
+}
+
+static void HR_Rewind(void *ctx) {
+  HybridIterator *hr = ctx;
+  hr->resultsPrepared = false;
+  if (hr->iter) VecSimQueryResult_IteratorFree(hr->iter);
+  if (hr->list) VecSimQueryResult_Free(hr->list);
+  // Clean the saved and returned results (in case of rewind).
+  while (heap_count(hr->topResults) > 0) {
```
Comment (DvirDukhan) [inline src/hybrid_reader.c:217]: will fail on normal KNN

File: tests/pytests/test_vecsim.py [hunk 18]
```diff
@@ -408,3 +389,187 @@ def test_memory_info(env):
...
+               'SORTBY', '__v_score',
+               'PARAMS', 2, 'vec_param', query_data.tobytes(),
+               'RETURN', 3, 't1', 't2', '__v_score').equal(expected_res_1)
+
+    # test with query attributes
+    env.expect('FT.SEARCH', 'idx', '(@t1|t2:(value text)=>{$inorder: true} @num:[10 30])=>[TOP_K 10 @v $vec_param]',
+               'SORTBY', '__v_score',
+               'WITHSCORES',
+               'PARAMS', 2, 'vec_param', query_data.tobytes(),
+               'RETURN', 2, 't1', 't2').equal([0L])
```
Comment (DvirDukhan) [inline tests/pytests/test_vecsim.py:575]: i i i 
empty line

File: src/hybrid_reader.c (hunk 13 above)
Comment (alonre24) [reply src/hybrid_reader.c:18]: I don't think so, do you?

File: src/hybrid_reader.c (hunk 16 above)
Comment (alonre24) [reply src/hybrid_reader.c:190]: no. I will add unit tests for the iterator...

File: src/hybrid_reader.c (hunk 17 above)
Comment (alonre24) [reply src/hybrid_reader.c:217]: fixing

**Review by DvirDukhan — APPROVED** [review]

**Review by ashtul — APPROVED** [review]
LGTM

**Pull request merged**